### ISSUE TITLE
Add more languages

### ISSUE THREAD

**Issue by qaq1000w** [Specy/genshin-music#40]
Support multiple languages,Such as Chinese

支持多种语言，如中文
感谢

**Comment by Specy** [comment]
It's on the list for the 4.0 release, but it's a huge effort and might take a while to do

**Comment by cangerjun** [comment]
> It's on the list for the 4.0 release, but it's a huge effort and might take a while to do

As long as you make the code you need translated publicly available, someone will come to complete the translation work.

**Comment by Specy** [comment]
> As long as you make the code you need translated publicly available, someone will come to complete the translation work.

Yes that is the issue, making the underlying code work with translations

**Comment by Specy** [comment]
I'll see tomorrow if I can plan it out and start working on it, it will take precedence over held instruments

**Comment by Specy** [comment]
Ok it took me something like 12 hours of work to extract all the text into a single file, if someone wants to help adding the Chinese translations (or any other translations that is), the file is here:
https://github.com/Specy/genshin-music/blob/Dev/src/i18n/locales/zh/index.ts
Just need to replace the right side of each string with it's translated counterpart, anything inside `{{ }}` doesn't have to be changed
example:
```js
//original
 delete_song: `Are you sure you want to delete the song "{{song_name}}"?`
//translated (in italian)
 delete_song: `Sei sicuro di voler cancellare la canzone "{{song_name}}"?`
```

**Comment by Specy** [comment]
@qaq1000w you can directly change the English one, I made a copy of the English one as reference to translate, @cangerjun seems to be working on the PR to add Chinese in #42,

**Comment by ptr2019** [comment]
> Something like this? 类似这样？ ![Stitch_20240511_085012](https://private-user-images.githubusercontent.com/80654352/329742377-aefe8d20-6a19-4579-8d49-94c750477a0a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU0MDY4NzcsIm5iZiI6MTcxNTQwNjU3NywicGF0aCI6Ii84MDY1NDM1Mi8zMjk3NDIzNzctYWVmZThkMjAtNmExOS00NTc5LThkNDktOTRjNzUwNDc3YTBhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTExVDA1NDkzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVhZTYyYWI2YzNkYjQ2NjQ1MjljNjM0YzAxZGQwYmQ2NWU1MWE0MjJhNGExNzA4N2M4NmMyNzAxZjE2YWNmMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HI7W8YF3h1C9j8syTnSI7Wb0Gh9kVHvZx_RUBDKOaiI)

It seems that a separate file is given for each language, which can be revised by translating the corresponding part. Nevertheless, this issue may not cause much trouble.

**Comment by cangerjun** [comment]
> > @qaq1000w你可以直接改英文的，我把英文的复制了一份作为参考翻译，@cangerjun似乎正在致力于在#42中添加中文的 PR ，
> 
> this is good

你把KEY的翻译写成语音高是不对的，key就是曲调的意思，键盘用简谱谢谢。

**Comment by cangerjun** [comment]
> > @qaq1000w你可以直接改英文的，我把英文的复制了一份作为参考翻译，@cangerjun似乎正在致力于在#42中添加中文的 PR ，
> 
> this is good

在所有用到key的句子你都翻译成了音高，音高一般是指按键的音高。
然后bpm我认为并不需要翻译，翻译后很不习惯（节拍数很少有人这样讲），

**Comment by cangerjun** [comment]
> > > > @qaq1000w你可以直接改英文的，我把英文的复制了一份作为参考翻译，@cangerjun似乎正在致力于在#42中添加中文的 PR ，
> > > 
> > > 
> > > this is good
> > 
> > 
> > 在所有用到key的句子你都翻译成了音高，音高一般是指按键的音高。 然后bpm我认为并不需要翻译，翻译后很不习惯（节拍数很少有人这样讲），
> 
> 我现在没有在弄了

已经合并你的了，我去邮箱中和@Specy 说一下好了

**Comment by Specy** [comment]
@cangerjun I've merged the changes and published it in the app, if you want to make changes to the translation go ahead, i'm not sure what i'd have to change. I'll also ask reviews of the translation to other chinese users to make sure everything is correct

**Comment by ptr2019** [comment]
> > > > > @qaq1000w你可以直接改英文的，我把英文的复制了一份作为参考翻译，@cangerjun似乎正在致力于在#42中添加中文的 PR ，
> > > > 
> > > > 
> > > > this is good
> > > 
> > > 
> > > 在所有用到key的句子你都翻译成了音高，音高一般是指按键的音高。 然后bpm我认为并不需要翻译，翻译后很不习惯（节拍数很少有人这样讲），
> > 
> > 
> > 我现在没有在弄了
> 
> 已经合并你的了，我去邮箱中和@Specy 说一下好了

有道理，这里key指的是调高（主音的音高）也好，还有bpm不用翻译也是。大佬有时间改一下可以嘛（这个是理论水平不够嘿嘿）。

**Comment by Specy** [comment]
If you find any issues in the translation or parts which were not translated, please let me know here

**Comment by Specy** [comment]
I added a few more things that need to be translated in [here](https://github.com/Specy/genshin-music/blob/Dev/src/i18n/locales/zh/index.ts), including instrument names. Could someone translate them? 

I also added the flags and put the language selection option in the settings of the composer and player

**Comment by cangerjun** [comment]
> I added a few more things that need to be translated in [here](https://github.com/Specy/genshin-music/blob/Dev/src/i18n/locales/zh/index.ts), including instrument names. Could someone translate them?
> 
> I also added the flags and put the language selection option in the settings of the composer and player

我已经完成翻译工作，而且在昨天的提交中我上传了一份README-ZH 文件，并且下原版README.me文件中添加了一段切换不同语言文件的代码。
I have completed the translation work, and in yesterday's submission, I uploaded a README-ZH file and added a code to switch between different language files in the original README.me file.

**Comment by Specy** [comment]
@cangerjun you had translated the instruments name and the other things that needed translation, but you closed the PR, there are still some thing left to translate, could someone finish those out? 

I'll publish it live after those are finished

**Comment by ptr2019** [comment]
> If you find any issues in the translation or parts which were not translated, please let me know here

It seems that the replacement is incomplete? 
The drop down menus of note type and instrument remain in English. 
![图片](https://github.com/Specy/genshin-music/assets/56548868/c46fe041-91c3-4e7d-a88b-762e37376d99)
![图片](https://github.com/Specy/genshin-music/assets/56548868/0cc696eb-46f7-4ef3-b58e-71db20793b7c)
What do you say?

**Comment by Specy** [comment]
Ah right I forgot the "note name type", I'll check later why the instrument aren't being translated

**Comment by Specy** [comment]
> > If you find any issues in the translation or parts which were not translated, please let me know here
> 
> It seems that the replacement is incomplete? The drop down menus of note type and instrument remain in English. ![图片]
What do you say?

Ok i managed to fix this, it would take a lot of more effort to support adding the translation to "note name type" so for now i'll keep it as it is, i've fixed the instruments translation though

**Comment by Specy** [comment]
I added those plus a few more in [the latest commit](https://github.com/Specy/genshin-music/commit/018e59cff6524ee6312b43008709cfb9c57a29b3#diff-c25b155cf9019f65bdb6c81b1e62fe3a0bfb3298788576b1844bbebe95a7c983), check that commit to know which lines were added and that still need translation

**Comment by ptr2019** [comment]
> > > If you find any issues in the translation or parts which were not translated, please let me know here
> > 
> > 
> > It seems that the replacement is incomplete? The drop down menus of note type and instrument remain in English. ![图片]
> > What do you say?
> 
> Ok i managed to fix this, it would take a lot of more effort to support adding the translation to "note name type" so for now i'll keep it as it is, i've fixed the instruments translation though

Got it.

> ![Screenshot_2024-05-19-00-05-18-134_com android chrome](https://private-user-images.githubusercontent.com/80654352/331814352-a22a7951-1caf-4fa6-bcd9-6dea503003f8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwNzI1OTAsIm5iZiI6MTcxNjA3MjI5MCwicGF0aCI6Ii84MDY1NDM1Mi8zMzE4MTQzNTItYTIyYTc5NTEtMWNhZi00ZmE2LWJjZDktNmRlYTUwMzAwM2Y4LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE4VDIyNDQ1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmMDNiMDc4OGIxZjU4NDVkMzQ3MjE5ZjE5ZWI2NjRiZDc5NjJjODc2NDYwODViMGM2NjcwZmUyNmFiZWMyM2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V_PET5AXMrlvL6HalLOcf7l4ufLbCiGlNPVDS96HI0Y) I think it's settled. @ptr2019

Yes, I see.



> I added those plus a few more in [the latest commit](https://github.com/Specy/genshin-music/commit/018e59cff6524ee6312b43008709cfb9c57a29b3#diff-c25b155cf9019f65bdb6c81b1e62fe3a0bfb3298788576b1844bbebe95a7c983), check that commit to know which lines were added and that still need translation
> 
> We are almost there!

Looking on it awa.

**Comment by Specy** [comment]
Ok so, everything seems to be done, i'll give this a few days to test everything out, if there are no issues i'll publish it

**Comment by Specy** [comment]
> Ok so, everything seems to be done, i'll give this a few days to test everything out, if there are no issues i'll publish it

Nevermind! There are a few more xD [look at the latest commit here](https://github.com/Specy/genshin-music/commit/837978b2ee3e72082d3707bc3c276231043b033a#diff-c25b155cf9019f65bdb6c81b1e62fe3a0bfb3298788576b1844bbebe95a7c983)

**Comment by Specy** [comment]
I have one more key to add:
```
how_to_install: "How to add to home screen" //opens the blog on how to install the app to home screen
```

**Comment by Specy** [comment]
I was suggested to change those keys here: https://github.com/Specy/genshin-music/commit/4bbd943ad74ee283e491dbad75ea6c1d2da663e9 are they correct? @ptr2019

**Comment by ptr2019** [comment]
Yes, some of the wrong translation are corrected ("编辑" for "edit", while "退出" for "exit"), while some of the specific terms in the game are reasonable (e.g., ”屁琴“ as nickname for 低音提琴/”contrabass“. I have consulted on this) .
![图片](https://github.com/Specy/genshin-music/assets/56548868/57e242a7-568b-4ad7-bac6-1ebf69d2670e)

I have also confirmed that ”变音记号“ is the correct term for "accidental". Again, sorry for the lack of professionalism in music composition area. 

To sum up, I have no objection on changing those keys.

**Comment by ptr2019** [comment]
Then ”连接“ is a general term appropriate to indicate "connect to something".

Glad to devote my efforts.  It is quite a nice experience~

**Comment by Specy** [comment]
I've merged the translations to the main branch in commit 15591be, I also published the V3.4 update live (includes the translations)

Thanks for everyone who helped translating!
